mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+#[allow(dead_code)]
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub(crate) struct FilePointers {
+    write: usize,
+    read: usize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around and cause us to lose data due to
+/// variable sized data being written. If W + data size > R and should write
+/// precede read, the data could become corrupted.
+///
+/// Either this is prevented by scanning the file for the next block and
+/// checking if the write will fit. Or, W + data must not pass both D and R.
+/// - Not passing D helps with guaranteeing all data read was handled.
+/// - Not passing R helps avoid corruption.
+#[allow(dead_code)]
+pub(crate) struct RingBuffer<'a> {
+    flush_options: FlushOptions,
+    flush_state: FlushState,
+    mmap: &'a mut MmapMut,
+    max_file_size: usize,
+    order: usize,
+    pointers: FilePointers,
+    waker: Option<Waker>,
+}
+
+#[allow(dead_code)]
+impl<'a> RingBuffer<'a> {
+    fn new(mmap: &'a mut MmapMut, max_file_size: usize, flush_options: FlushOptions) -> Self {
+        let file_pointers =
+            find_pointers_post_crash(&mmap, max_file_size).expect("Failed to init file pointers");
+        Self {
+            flush_options,
+            flush_state: FlushState::new(),
+            mmap,
+            max_file_size,
+            order: usize::default(),
+            pointers: file_pointers,
+            waker: None,
+        }
+    }
+
+    fn insert(&mut self, publication: &Publication) -> StorageResult<Key> {
+        let timer = Instant::now();
+        let data = binary_serialize(publication)?;
+        let data = Data::new(&data);
+        let data_size = binary_serialize_size(&data)?;
+        let data_hash = calculate_hash(&data);
+
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let order = self.order;
+        let key = write_index;
+
+        let block_header =
+            BlockHeaderWithHash::new(data_hash, data_size, order, read_index, write_index);
+        let block_size = serialized_block_size()?;
+
+        let total_size = block_size + data_size;
+        // Check to see if we might corrupt data if we write, if so, return err of full.
+        // There are two cases for this:
+        // 1. write has wrapped around and is now behind read
+        if write_index < read_index && write_index + total_size > read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::Full));
+        }
+        // 2. write has reached the end (or close enough) but read has not moved
+        if read_index == 0 && write_index + total_size > self.max_file_size {
```

Comment:
I think this needs to be checked for read_index other than 0 as well... something like this
```rust
        if write_index > read_index
            && write_index + total_size > self.max_file_size
            && write_index + total_size % self.max_file_size > read_index
        {
            return Err(StorageError::RingBuffer(RingBufferError::Full));
        }
```

Reply:
Not sure on the above situation here... The code allows for a wrap around.
Imagine the '-' is just random garbage data and the '/' is a block and data.
[//:-------://]

We support this situation where we have some free space at the beginning and the end, but cannot fit entirely at the end. So, the `write_index + total_size > self.max_file_size` case seems odd to me. But the second part you suggest `write_index + total_size % self.max_file_size > read_index` should probably just be the first check I have i.e:
```rust
 if write_index < read_index && write_index + total_size > read_index {
            return Err(StorageError::RingBuffer(RingBufferError::Full));
 }
```
should be
```rust
 if write_index + total_size % self.max_file_size > read_index{
            return Err(StorageError::RingBuffer(RingBufferError::Full));
 }
```